### PR TITLE
.defineSubnet() support for virtual network; also CRUD for NetworkSecurityGroup

### PR DESCRIPTION
**Pull request by martinsawicki**
(still todo: NSG association, when NSGs are supported)

### CODE REVIEW

File: azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/NetworkImpl.java [hunk 1]
```diff
@@ -160,18 +160,19 @@ public NetworkImpl withAddressSpace(String cidr) {
...
             this.withAddressSpace("10.0.0.0/16");
         }
 
         // Create a subnet as needed, covering the entire first address space
-        // TODO: this shouldn't happen during Update -- may need to move to initializer or something...
-        if(this.inner().subnets().size() == 0) {
+        // Future: this shouldn't happen during Update -- may need to move to initializer or something...
+        if (this.inner().subnets().size() == 0) {
             this.withSubnet("subnet1", this.addressSpaces().get(0));
         }
```
Comment (anuchandy) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/NetworkImpl.java]: you can use isInCreateMode() method here to ensure this happens only during create

```java
if (isInCreateMode() && this.inner().subnets().size() == 0) {
   this.withSubnet("subnet1", this.addressSpaces().get(0));
}
```

File: azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/NetworkImpl.java [hunk 2]
```diff
@@ -112,17 +108,17 @@ public NetworkImpl withSubnets(Map<String, String> nameCidrPairs) {
...
 
         // Remove from inner
         List<SubnetInner> innerSubnets = this.inner().subnets();
-        for(int i=0; i<innerSubnets.size(); i++) {
-            if(innerSubnets.get(i).name().equalsIgnoreCase(name)) {
+        for (int i = 0; i < innerSubnets.size(); i++) {
+            if (innerSubnets.get(i).name().equalsIgnoreCase(name)) {
                 innerSubnets.remove(i);
                 break;
             }
```
Comment (anuchandy) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/NetworkImpl.java]: Should we just remove only from List<Subnet> collection? When we are ready to build payload (in `createResource`) we could iterate though the List<Subnet> and populate the inner.subnets.

File: azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/NetworkImpl.java [hunk 3]
```diff
@@ -137,10 +133,14 @@ public NetworkImpl withAddressSpace(String cidr) {
         return this;
     }
 
+    @Override
+    public SubnetImpl defineSubnet(String name) {
+        SubnetInner inner = new SubnetInner();
+        inner.setName(name);
+        return new SubnetImpl(name, inner, this);
```
Comment (anuchandy) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/NetworkImpl.java]: related to previous comment.. we could add SubnetImpl to List<Subnet> here that ensure this list contains all the subnets when we populate the subnets in the payload in `createResource` method.

File: azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/NetworkImpl.java (hunk 2 above)
Comment (martinsawicki) [inline azure-mgmt-network/src/main/java/com/microsoft/azure/management/network/implementation/NetworkImpl.java]: when we separate the logic of removing from the "truth" (inner) from the logic of removing from the cache, I worry it's an open invitation for bugs where the cache is not in synch with the "truth"...

**Pull request merged**